sql/core/src/main/scala/org/apache/spark/sql/execution/bucketing/CoalesceBucketInJoin.scala

```diff
@@ -0,0 +1,110 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.bucketing
+
+import org.apache.spark.sql.catalyst.catalog.BucketSpec
+import org.apache.spark.sql.catalyst.plans.Inner
+import org.apache.spark.sql.catalyst.plans.logical.{Filter, Join, LogicalPlan, Project}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.datasources.{HadoopFsRelation, LogicalRelation}
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * This rule injects a hint if one side of two bucketed tables can be coalesced
+ * when the two bucketed tables are inner-joined and they differ in the number of buckets.
+ */
+object CoalesceBucketInJoin extends Rule[LogicalPlan]  {
+  val JOIN_HINT_COALESCED_NUM_BUCKETS: String = "JoinHintCoalescedNumBuckets"
+
+  private val sqlConf = SQLConf.get
+
+  private def isPlanEligible(plan: LogicalPlan): Boolean = {
+    def forall(plan: LogicalPlan)(p: LogicalPlan => Boolean): Boolean = {
+      p(plan) && plan.children.forall(forall(_)(p))
+    }
+
+    forall(plan) {
+      case _: Filter | _: Project | _: LogicalRelation => true
+      case _ => false
+    }
+  }
+
+  private def getBucketSpec(plan: LogicalPlan): Option[BucketSpec] = {
+    if (isPlanEligible(plan)) {
+      plan.collectFirst {
+        case _ @ LogicalRelation(r: HadoopFsRelation, _, _, _)
+          if r.bucketSpec.nonEmpty && !r.options.contains(JOIN_HINT_COALESCED_NUM_BUCKETS) =>
+          r.bucketSpec.get
+      }
+    } else {
+      None
+    }
+  }
+
+  private def mayCoalesce(numBuckets1: Int, numBuckets2: Int): Option[Int] = {
+    assert(numBuckets1 != numBuckets2)
+    val (small, large) = (math.min(numBuckets1, numBuckets2), math.max(numBuckets1, numBuckets2))
+    // A bucket can be coalesced only if the bigger number of buckets is divisible by the smaller
+    // number of buckets because bucket id is calculated by modding the total number of buckets.
+    if ((large % small == 0) &&
+      (large - small) <= sqlConf.getConf(SQLConf.COALESCE_BUCKET_IN_JOIN_MAX_NUM_BUCKETS_DIFF)) {
+      Some(small)
+    } else {
+      None
+    }
+  }
+
+  private def addBucketHint(plan: LogicalPlan, hint: (String, String)): LogicalPlan = {
+    plan.transformUp {
+      case l @ LogicalRelation(r: HadoopFsRelation, _, _, _) =>
+        l.copy(relation = r.copy(options = r.options + hint)(r.sparkSession))
```

Comment:
hm, I still feel a bit weird about using options for this purpose. With the current approach, it seems users can arbitrarily set a value in `JOIN_HINT_COALESCED_NUM_BUCKETS` via datasource options.

Reply:
Ah, good point. But I am not sure if I can do something similar in a strategy; for example, if I see a `Join` in a strategy, I need to act on it, and if I see a scan node, I cannot traverse up the `TreeNode`. Modifying `BucketSpec` is not desirable either, right?

Reply:
> for example, if I see a Join in a strategy, I need to act on it, and if I see a scan node, I cannot traverse up the TreeNode


You cannot extract a join plan pattern, e.g., `Join(LogicalRelation(HadoopFsRelation), LogicalRelation(HadoopFsRelation))`,  just like `ScanOperation`?

> Modifying BucketSpec is not desirable either, right?

Yea, `BucketSpec` is an item in catalog, so I think we cannot put optimization hints in it.

Reply:
> You cannot extract a join plan pattern, e.g., `Join(LogicalRelation(HadoopFsRelation), LogicalRelation(HadoopFsRelation))`, just like `ScanOperation`?

If I match a `Join` in `FileSourceStrategy`, I need to return `SparkPlan` for `Join`, no?

Reply:
How about using `PlanLater` there?

Reply:
There are the options I can think of:
1. Handle if `JOIN_HINT_COALESCED_NUM_BUCKETS` is incorrectly set.
2. Update `HadoopFsRelation` to have an extra field.
3. Update `LogicalRelation` to have a `hint` similar to `Join`.
What do you think?

Reply:
Or, how about adding a wrapper logical plan? e.g., `Join(LogicalRelation, CoalesceBuckets(LogicalRelation))`, then we put `#coalesced buckets` in `CoalesceBuckets`. In the strategy, we replate `CoalesceBuckets(LogicalRelation)` with `FileSourceScanExec`.

Reply:
Thanks, that should work!